apis/Google.Cloud.AspNetCore.DataProtection.SecretManager/Google.Cloud.AspNetCore.DataProtection.SecretManager/CloudSecretManagerXmlRepository.cs

```diff
@@ -0,0 +1,120 @@
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.ResourceNames;
+using Google.Cloud.SecretManager.V1;
+using Google.Protobuf;
+using Grpc.Core;
+using Microsoft.AspNetCore.DataProtection.Repositories;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Xml.Linq;
+using Secret = Google.Cloud.SecretManager.V1.Secret;
+
+namespace Google.Cloud.AspNetCore.DataProtection.SecretManager;
+
+/// <summary>
+/// Implementation of <see cref="IXmlRepository"/> that stores the protected elements in a Google Cloud Secret Manager.
+/// This class is configured by <see cref="GoogleCloudDataProtectionBuilderExtensions.PersistKeysToGoogleCloudSecretManager(Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder, string, string)"/>
+/// (and other overloads).
+/// </summary>
+internal class CloudSecretManagerXmlRepository : IXmlRepository
+{
+    private readonly SecretManagerServiceClient _secretManagerServiceClient;
+    private readonly SecretName _secretName;
+    private readonly ProjectName _projectName;
+    private readonly SecretVersionName _version;
+
+    internal CloudSecretManagerXmlRepository(SecretManagerServiceClient client, string secretName, string projectId)
+    {
+        _secretManagerServiceClient = GaxPreconditions.CheckNotNull(client, nameof(client));
+        GaxPreconditions.CheckNotNull(secretName, nameof(secretName));
+        GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+        _projectName = new ProjectName(projectId);
+        _secretName = new SecretName(projectId, secretName);
+
+        // Version is populated in such a way that it always refers to the most latest version.
+        // For more information on version naming refer to
+        // https://cloud.google.com/dotnet/docs/reference/Google.Cloud.SecretManager.V1/latest/Google.Cloud.SecretManager.V1.AccessSecretVersionRequest#Google_Cloud_SecretManager_V1_AccessSecretVersionRequest_Name
+        _version = SecretVersionName.FromProjectSecretSecretVersion(projectId, secretName, "latest");
+    }
+
+    /// <inheritdoc />
+    public IReadOnlyCollection<XElement> GetAllElements()
+    {
+        if (!SecretExists())
+        {
+            return Array.Empty<XElement>();
+        }
+        var accessSecretVersionResponse = _secretManagerServiceClient.AccessSecretVersion(_version);
+
+        if (accessSecretVersionResponse != null)
+        {
+            var secretPayload = accessSecretVersionResponse.Payload.Data.ToStringUtf8();
+            XDocument document = XDocument.Parse(secretPayload);
+            var xElements = document.Root.Elements().ToList();
+            return xElements;
+        }
+        return Array.Empty<XElement>();
+    }
+
+    /// <inheritdoc />
+    public void StoreElement(XElement element, string friendlyName)
+    {
+        if (!SecretExists())
```

Comment:
There's a race condition here if multiple clients try to do this at the same time.
Maybe it's worth just unconditionally creating the secret, and catching the specific "it already exists" failure?
If we're worried that's expensive, we could check for existence, then try to create the secret but *still* catch the "it already exists" failure.

@amanda-tarafa thoughts welcome here.

Reply:
updated the logic Please have a look and share your thoughts